README


This archive contains the distribution Email-Send,
version 2.199:

  Simply Sending Email

This software is copyright (c) 2004 by Casey West.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
2.04    2006-01-28

  - Integrated Email::Send::Test (Adam Kennedy).

2.03    2006-01-28

  - Updated tests to only check the interface of distributed mailers.
  - Updated Email::Send::Sendmail to provide much more detail when mail fails
    to send. (Dave Rolksy and Casey West).

2.01    2006-01-17

  - Updated tests to only check distributed mailers.
  - Updated version numbers in all packages for proper indexing.

2.00    2005-12-15

  - Grafted Classic interface but did not document it.

1.99_01    2005-05-09

  - Major changes to API.
  - Changes to Makefile.PL to aid in recommended module builds.
  - Alpha-release of new interface.

1.46   2005-01-08

  - SMTP was broken. I made a horrible mistake with my use of eval.
    Honestly, I feel quite silly. I'm sorry.
  - Added eg/send-email.pl, a program to aid in testing and debuging
    of this code.

1.45   2004-12-22

  - Bug in Email::Send::SMTP in collecting addresses to send for RCPT.
  - Added more error checking in Email::Send::SMTP to return proper
    diagnostics.
  - Updated dependencies.

1.44   2004-12-17

  - Make Email::Send::SMTP use Return::Value for its returns. These
    are compatible with previous behavior, but offer more detail on
    just what went wrong.
  - Require Return::Value as a dependency. Other mailers may begin
    using it for returns in the future.

1.43   2004-08-07

  - Make Email::Send::SMTP send args to Net::SMTP with host first.
    This makes older Net::SMTPs happy.

1.42   2004-07-29

  - Remove hard dependency on Email::Abstract (for IO::All).
  - Declare hard dependency on Email::Simple.

1.41   2004-07-20

  - Only set defaults for $Sendmail and $Qmail in those mailers
    when the package variable is not already set (Peter Scott).
  - SSL support for Email::Send::SMTP.
  - Use different version scheme, I need more numbers.

1.4    2004-07-08

  - Relaxed prereq on IO::All for testing.
  - Included Bcc in list of addresses to MAIL TO: in
    Email::Send::SMTP.
  - Now supports Net::SMTP's auth() (Aaron Trevena).

1.3    2004-07-05

  - Doc typo in Email::Send::Sendmail (Martin 'Kingpin' Thurn).

1.2    2004-06-29

  - Email::Send::SMTP now sends multiple MAIL commands for
    each address in the To and Cc headers. It does not fail
    when it encounters bad addresses, as long as at least one
    address succeeds.

1.1    2004-05-27

  - Initial version.


lib/Email/Send/IO.pm
package Email::Send::IO;
# $Id: IO.pm,v 1.6 2006/01/28 23:02:44 cwest Exp $
use strict;

use UNIVERSAL::require;
use Return::Value;

use vars qw[$VERSION];
$VERSION   = '2.04';

use vars qw[@IO];
@IO      = ('=') unless @IO;

sub is_available {
    return   IO::All->require
           ? success
           : failure $UNIVERSAL::require::ERROR;
}

sub send {
    my ($class, $message, @args) = @_;
    IO::All->require or return failure;
    IO::All->import;
    @args = (@IO) unless @args;
    eval {io(@args)->print($message->as_string)};
    return failure $@ if $@;
    return success;
}

1;

__END__

=head1 NAME

Email::Send::IO - Send messages using IO operations

=head1 SYNOPSIS

  use Email::Send;

  my $mailer = Email::Send->new({mailer => 'IO'});

  $mailer->send($message); # To STDERR

  $mailer->mailer_args('filename.txt');
  $mailer->send($message); # write file

=head1 DESCRIPTION

This is a mailer for C<Email::Send> that will send a message using IO
operations. By default it sends mail to STDERR, very useful for debugging.
The IO functionality is built upon C<IO::All>. Any additional arguments
passed to C<send> will be used as arguments to C<IO::All::io>.

You can globally change where IO is sent by modifying the C<@Email::Send::IO::IO>
package variable.

  @Email::Send::IO::IO = ('-'); # always append to STDOUT.

=head2 Examples

Sending to STDOUT.

  send IO => $message, '-';

Send to a socket.

  send IO => $message, 'server:1337';

=head1 SEE ALSO

L<Email::Send>,
L<IO::All>,
L<perl>.

=head1 AUTHOR

Casey West, <F<casey@geeknest.com>>.

=head1 COPYRIGHT

  Copyright (c) 2005 Casey West.  All rights reserved.
  This module is free software; you can redistribute it and/or modify it
  under the same terms as Perl itself.

=cut


lib/Email/Send/NNTP.pm
package Email::Send::NNTP;
# $Id: NNTP.pm,v 1.5 2006/01/28 23:02:44 cwest Exp $
use strict;

use vars qw[$NNTP $VERSION];
use Net::NNTP;
use Return::Value;
use UNIVERSAL::require;

$VERSION   = '2.04';

sub is_available {
    return   Net::NNTP->require
           ? success
           : failure;
}

sub send {
    my ($class, $message, @args) = @_;
    Net::NNTP->require;
    if ( @_ > 1 ) {
        $NNTP->quit if $NNTP;
        $NNTP = Net::NNTP->new(@args);
        return failure unless $NNTP;
    }
    return failure unless $NNTP->post( $message->as_string );
    return success;
}

sub DESTROY {
    $NNTP->quit if $NNTP;
}

1;

__END__

=head1 NAME

Email::Send::NNTP - Post Messages to a News Server

=head1 SYNOPSIS

  use Email::Send;

  my $mailer = Email::Send->new({mailer => 'NNTP'});
  
  $mailer->mailer_args([Host => 'nntp.example.com']);
  
  $mailer->send($message);

=head1 DESCRIPTION

This is a mailer for C<Email::Send> that will post a message to a news server.
The message must be formatted properly for posting. Namely, it must contain a
I<Newsgroups:> header. At least the first invocation of C<send> requires
a news server arguments. After the first declaration the news server will
be remembered until such time as you pass another one in.

=head1 SEE ALSO

L<Email::Send>,
L<Net::NNTP>,
L<perl>.

=head1 AUTHOR

Casey West, <F<casey@geeknest.com>>.

=head1 COPYRIGHT

  Copyright (c) 2004 Casey West.  All rights reserved.
  This module is free software; you can redistribute it and/or modify it
  under the same terms as Perl itself.

=cut


lib/Email/Send/SMTP.pm
package Email::Send::SMTP;
# $Id: SMTP.pm,v 1.12 2006/01/28 23:02:44 cwest Exp $
use strict;

use vars qw[$SMTP $VERSION];
use Net::SMTP;
use Email::Address;
use Return::Value;
use UNIVERSAL::require;

$VERSION   = '2.04';

sub is_available {
    my ($class, %args) = @_;
    my $success = 1;
    $success = Net::SMTP->require;
    $success = Net::SMTP::SSL->require if $args{ssl};
    return   $success
           ? success
           : failure;
}

sub send {
    my ($class, $message, @args) = @_;
    if ( @_ > 1 ) {
        my %args;
        if ( @args % 2 ) {
            my $host = shift @args;
            %args = @args;
            $args{Host} = $host;
        } else {
            %args = @args;
        }

        my $host = delete($args{Host}) || 'localhost';
        if ( $args{ssl} ) {
            require Net::SMTP::SSL;
            $SMTP->quit if $SMTP;
            $SMTP = Net::SMTP::SSL->new($host, %args);
            return failure "Couldn't connect to $host" unless $SMTP;
        } else {
            $SMTP->quit if $SMTP;
            $SMTP = Net::SMTP->new($host, %args);
            return failure "Couldn't connect to $host" unless $SMTP;
        }
        
        my ($user, $pass)
          = @args{qw[username password]};
        if ( $user ) {
            $SMTP->auth($user, $pass)
              or return failure "Couldn't authenticate '$user:...'";
        }
    }
    
    my @bad;
    my @to;
    eval {
        my $from =
          (Email::Address->parse($message->header('From')))[0]->address;
        $SMTP->mail($from) or return failure "FROM: <$from> denied";

        @to = map {
                   map { $_->address }
                       Email::Address->parse($message->header($_))
                  } qw[To Cc Bcc];
        my @ok = $SMTP->to(@to, { SkipBad => 1 });

        if ( @to != @ok ) {
            my %to; @to{@to} = (1) x @to;
            delete @to{@ok};
            @bad = keys %to;
        }
    };
    return failure $@ if $@;

    return failure "No valid recipients" if @bad == @to;

    return failure "Can't send data"
      unless $SMTP->data( $message->as_string );

    return success "Message sent", prop => {
        bad => [ @bad ],
    };
}

sub DESTROY {
    $SMTP->quit if $SMTP;
}

1;

__END__

=head1 NAME

Email::Send::SMTP - Send Messages using SMTP

=head1 SYNOPSIS

  use Email::Send;

  my $mailer = Email::Send->new({mailer => 'SMTP'});
  
  $mailer->mailer_args([Host => 'smtp.example.com:465', ssl => 1])
    if $USE_SSL;
  
  $mailer->send($message);

=head1 DESCRIPTION

This mailer for C<Email::Send> uses C<Net::SMTP> to send a message with
an SMTP server. The first invocation of C<send> requires an SMTP server
arguments. Subsequent calls will remember the the first setting until
it is reset.

Any arguments passed to C<send> will be passed to C<< Net::SMTP->new() >>,
with some exceptions. C<username> and C<password>, if passed, are
used to invoke C<< Net::SMTP->auth() >> for SASL authentication support.
C<ssl>, if set to true, turns on SSL support by using C<Net::SMTP::SSL>.

SMTP can fail for a number of reasons. All return values from this
package are true or false. If false, sending has failed. If true,
send succeeded. The return values are C<Return::Value> objects, however,
and contain more information on just what went wrong.

Here is an example of dealing with failure.

  my $return = send SMTP => $message, 'localhost';
  
  die "$return" if ! $return;

The stringified version of the return value will have the text of the
error. In a conditional, a failure will evaluate to false.

Here's an example of dealing with success. It is the case that some
email addresses may not succeed but others will. In this case, the
return value's C<bad> property is set to a list of bad addresses.

  my $return = send SMTP => $message, 'localhost';

  if ( $return ) {
      my @bad = @{ $return->prop('bad') };
      warn "Failed to send to: " . join ', ', @bad
        if @bad;
  }

For more information on these return values, see L<Return::Value>.

=head1 SEE ALSO

L<Email::Send>,
L<Net::SMTP>,
L<Net::SMTP::SSL>,
L<Email::Address>,
L<Return::Value>,
L<perl>.

=head1 AUTHOR

Casey West, <F<casey@geeknest.com>>.

=head1 COPYRIGHT

  Copyright (c) 2004 Casey West.  All rights reserved.
  This module is free software; you can redistribute it and/or modify it
  under the same terms as Perl itself.

=cut


lib/Email/Send/Qmail.pm
package Email::Send::Qmail;
# $Id: Qmail.pm,v 1.5 2006/01/28 21:44:17 cwest Exp $
use strict;

use vars qw[$QMAIL $VERSION];
$QMAIL   ||= q[qmail-inject];

use Return::Value;

$VERSION   = '2.03';

sub is_available {
    return   `which $QMAIL`
           ? success
           : failure;
}

sub send {
    my ($class, $message, @args) = @_;
    open QMAIL, "| $QMAIL @args" or return failure;
    print QMAIL $message->as_string;
    close QMAIL;
    return success;
}

1;

__END__

=head1 NAME

Email::Send::Qmail - Send Messages using qmail-inject

=head1 SYNOPSIS

  use Email::Send;

  Email::Send->new({mailer => 'Qmail'})->send($message);

=head1 DESCRIPTION

This mailer for C<Email::Send> uses C<qmail-inject> to put a message in
the Qmail spool. It I<does not> try hard to find the executable. It just
calls C<qmail-inject> and expects it to be in your path. If that's not
the case, or you want to explicitly define the location of your
executable, alter the C<$Email::Send::Qmail::QMAIL> package variable.

  $Email::Send::Qmail::QMAIL = '/usr/sbin/qmail-inject';

Any arguments passed to C<send> will be passed to C<qmail-inject>.

=head1 SEE ALSO

L<Email::Send>,
L<perl>.

=head1 AUTHOR

Casey West, <F<casey@geeknest.com>>.

=head1 COPYRIGHT

  Copyright (c) 2004 Casey West.  All rights reserved.
  This module is free software; you can redistribute it and/or modify it
  under the same terms as Perl itself.

=cut


lib/Email/Send/Sendmail.pm
package Email::Send::Sendmail;
# $Id: Sendmail.pm,v 1.6 2006/01/28 21:44:18 cwest Exp $
use strict;

use Return::Value;

use vars qw[$SENDMAIL $VERSION];

$VERSION   = '2.03';

sub is_available {
    my $class = shift;
    my $status = '';
       $status = "No Sendmail found" unless $class->_find_sendmail;
    return success $status;
}

sub _find_sendmail {
    my $class = shift;
    return $SENDMAIL if defined $SENDMAIL;

    my @path = split /:/, $ENV{PATH};
    my $sendmail;
    for (@path) {
        if ( -x "$_/sendmail" ) {
            $sendmail = "$_/sendmail";
            last;
        }
    }
    return $sendmail;
}

sub send {
    my ($class, $message, @args) = @_;
    my $mailer = $class->_find_sendmail;
    return failure "Found $mailer but cannot execute it"
        unless -x $mailer;
    open SENDMAIL, "| $mailer -t -oi @args"
        or return failure "Error executing $mailer: $!";
    print SENDMAIL $message->as_string
        or return failure "Error printing via pipe to $mailer: $!";
    close SENDMAIL;
    return success;
}

1;

__END__

=head1 NAME

Email::Send::Sendmail - Send Messages using sendmail

=head1 SYNOPSIS

  use Email::Send;

  Email::Send->new({mailer => 'Sendmail'})->send($message);

=head1 DESCRIPTION

This mailer for C<Email::Send> uses C<sendmail> to send a message. It
I<does not> try hard to find the executable. It just calls
C<sendmail> and expects it to be in your path. If that's not the
case, or you want to explicitly define the location of your executable,
alter the C<$Email::Send::Sendmail::SENDMAIL> package variable.

  $Email::Send::Sendmail::SENDMAIL = '/usr/sbin/sendmail';

Any arguments passed to C<send> will be passed to C<sendmail>. The
C<-t -oi> arguments are sent automatically.

=head1 SEE ALSO

L<Email::Send>,
L<perl>.

=head1 AUTHOR

Casey West, <F<casey@geeknest.com>>.

=head1 COPYRIGHT

  Copyright (c) 2004 Casey West.  All rights reserved.
  This module is free software; you can redistribute it and/or modify it
  under the same terms as Perl itself.

=cut


lib/Email/Send.pm
package Email::Send;
# $Id: Send.pm,v 1.16 2006/01/28 23:02:44 cwest Exp $
use strict;

use vars qw[$VERSION];
$VERSION   = '2.04';

use base qw[Class::Accessor::Fast];
use Email::Simple;
use Module::Pluggable search_path => 'Email::Send';
use UNIVERSAL::require;
use Return::Value;

=head1 NAME

Email::Send - Simply Sending Email

=head1 SYNOPSIS

  use Email::Send;

  my $message = <<'__MESSAGE__';
  To: recipient@example.com
  From: sender@example.com
  Subject: Hello there folks
  
  How are you? Enjoy!
  __MESSAGE__

  my $sender = Email::Send->new({mailer => 'SMTP'});
  $sender->mailer_args([Host => 'smtp.example.com']);
  $sender->send($message);
  
  # more complex
  my $bulk = Email::Send->new;
  for ( qw[SMTP Sendmail Qmail] ) {
      $bulk->mailer($_) and last if $bulk->mailer_available($_);
  }

  $bulk->message_modifier(sub {
      my ($sender, $message, $to) = @_;
      $message->header_set(To => qq[$to\@geeknest.com])
  });
  
  my @to = qw[casey chastity evelina casey_jr marshall];
  my $rv = $bulk->send($message, $_) for @to;

=head1 DESCRIPTION

This module provides a very simple, very clean, very specific interface
to multiple Email mailers. The goal if this software is to be small
and simple, easy to use, and easy to extend.

=head2 Constructors

=over 4

=item new()

  my $mailer = Email::Send->new({
      mailer      => 'NNTP',
      mailer_args => [ Host => 'nntp.example.com' ],
  });

Create a new mailer object. This method can take parameters for any of the data
properties of this module. Those data properties, which have their own accessors,
are listed under L<"Properties">.

=back

=head2 Properties

=over 4

=item mailer

The mailing system you'd like to use for sending messages with this object.
This is not defined by default. If you don't specify a mailer, all available
plugins will be tried when the C<send()> method is called until one succeeds.

=item mailer_args

Arguments passed into the mailing system you're using.

=item message_modifier

If defined, this callback is invoked every time the C<send()> method is called
on an object. The mailer object will be passed as the first argument. Second,
the actual C<Email::Simple> object for a message will be passed. Finally, any
additional arguments passed to C<send()> will be passed to this method in the
order they were recieved.

This is useful if you are sending in bulk.

=back

=cut

sub new {
    my ($class, $args) = @_;
    $args->{mailer_args} ||= [];
    my %plugins = map {
	                   my ($short_name) = /^Email::Send::(.+)/;
		               ($short_name, $_);
		              } $class->plugins;
	$args->{_plugin_list} = \%plugins;
    return $class->SUPER::new($args);
}
__PACKAGE__->mk_accessors(qw[mailer mailer_args message_modifier _plugin_list]);

=head2 Methods

=over 4

=item send()

  my $result = $mailer->send($message, @modifier_args);

Send a message using the predetermined mailer and mailer arguments. If you
have defined a C<message_modifier> it will be called prior to sending.

The first argument you pass to send is an email message. It must be in some
format that C<Email::Abstract> can understand. If you don't have C<Email::Abstract>
installed then sending as plain text or an C<Email::Simple> object will do.

Any remaining arguments will be passed directly into your defined
C<message_modifier>.

=cut

sub send {
    my ($self, $message, @args) = @_;
    my $simple = $self->_objectify_message($message);
    return failure "No message found." unless $simple;

	$self->message_modifier(
		$self, $simple,
		@args,
	) if $self->message_modifier;

	if ( $self->mailer ) {
		return $self->_send_it($self->mailer, $simple);
	}

	return $self->_try_all($simple);
}

=item all_mailers()

  my @available = $mailer->all_mailers;

Returns a list of availabe mailers. These are mailers that are
installed on your computer and register themselves as available.

=cut

sub all_mailers {
	my ($self) = @_;
	my @mailers;
	for ( keys %{$self->_plugin_list} ) {
		push @mailers, $_ if $self->mailer_available($_);
	}
	return @mailers;
}

=item mailer_available()

  # is SMTP over SSL avaialble?
  $mailer->mailer('SMTP')
    if $mailer->mailer_available('SMTP', ssl => 1);

Given the name of a mailer, such as C<SMTP>, determine if it is
available. Any additional arguments passed to this method are passed
directly to the C<is_available()> method of the mailer being queried.

=back

=cut

sub mailer_available {
	my ($self, $mailer, @args) = @_;
	if ( my $package = $self->_plugin_list->{$mailer} ) {
	    $package->require or return failure;
	    $package->can('is_available')
	      or return failure "Mailer $mailer doesn't report availability.";
		my $test = $package->is_available(@args);
		return $test unless $test;
		return success;
	}
	return failure "Mailer $mailer not found.";
}

sub _objectify_message {
    my ($self, $message) = @_;

    return $message if UNIVERSAL::isa($message, 'Email::Simple');
    return Email::Simple->new($message) unless ref($message);
    return Email::Abstract->cast($message => 'Email::Simple')
      if Email::Abstract->require;
    return undef;
}

sub _send_it {
	my ($self, $mailer, $message) = @_;
	my $test = $self->mailer_available($mailer);
	return $test unless $test;

    my $package = $self->_plugin_list->{$mailer};
    $package->require or return failure;
	return $package->send($message, @{$self->mailer_args});
}

sub _try_all {
	my ($self, $simple) = @_;
	foreach ( $self->all_mailers ) {
		my $sent = $self->_send_it($_, $simple);
		return $sent if $sent;
	}
	return failure "Unable to send message.";
}

# Classic Interface.

sub import {
    no strict 'refs';
    *{(caller)[0] . '::send'} = __PACKAGE__->can('_send_function');
}

sub _send_function {
    my ($mailer, $message, @args) = @_;
    __PACKAGE__->new({
    	mailer => $mailer,
    	mailer_args => \@args,
	})->send($message);
}

1;

__END__

=head2 Writing Mailers

  package Email::Send::Example;

  use UNIVERSAL::require;

  sub is_available {
      Net::Example->require;
  }

  sub send {
      my ($class, $message, @args) = @_;
      Net::Example->require;
      Net::Example->do_it($message) or return;
  }
  
  1;

Writing new mailers is very simple. If you want to use a short name
when calling C<send>, name your mailer under the C<Email::Send> namespace.
If you don't, the full name will have to be used. A mailer only needs
to implement a single function, C<send>. It will be called from
C<Email::Send> exactly like this.

  Your::Sending::Package::send($message, @args);

C<$message> is an Email::Simple object, C<@args> are the extra
arguments passed into C<Email::Send::send>.

Here's an example of a mailer that sends email to a URL.

  package Email::Send::HTTP::Post;
  use strict;

  use vars qw[$AGENT $URL $FIELD];
  use UNIVERSAL::require;
  use Return::Value;
  
  sub is_available {
	  LWP::UserAgent->require;
  }

  sub send {
      my ($message, @args);

	  LWP::UserAgent->require;

      if ( @args ) {
          my ($URL, $FIELD) = @args;
          $AGENT = LWP::UserAgent->new;
      }
      return failure "Can't send to URL if no URL and field are named"
        unless $URL && $FIELD;
      $AGENT->post($URL => { $FIELD => $message->as_string });
	  return success;
  }

  1;

This example will keep a UserAgent singleton unless new arguments are
passed to C<send>. It is used by calling C<Email::Send::send>.

  my $mailer = Email::Send->new({ mailer => 'HTTP::Post' });
  
  $mailer->mailer_args([ 'http://example.com/incoming', 'message' ]);

  $mailer->send($message);
  $mailer->send($message2); # uses saved $URL and $FIELD

=head1 SEE ALSO

L<Email::Simple>,
L<Email::Abstract>,
L<Email::Send::IO>,
L<Email::Send::NNTP>,
L<Email::Send::Qmail>,
L<Email::Send::SMTP>,
L<Email::Send::Sendmail>,
L<perl>.

=head1 AUTHOR

Casey West, <F<casey@geeknest.com>>.

=head1 COPYRIGHT

  Copyright (c) 2005 Casey West.  All rights reserved.
  This module is free software; you can redistribute it and/or modify it
  under the same terms as Perl itself.

=cut
